### PR TITLE
Optimise span_lines_str.

### PR DESCRIPTION
**Pull request by ltratt**
Currently this is O(n) -- this PR reduces it to O(log_n).

### CODE REVIEW

File: lrlex/src/lib/lexer.rs [hunk 1]
```diff
@@ -310,30 +310,18 @@ impl<'lexer, 'input: 'lexer, StorageT: Copy + Eq + Hash + PrimInt + Unsigned>
...
-        let (st, _) = surrounding_line_off(self, span.start());
-        let (_, en) = surrounding_line_off(self, span.end());
+        let (st, _) = match self.newlines.binary_search(&span.start()) {
+            Ok(j) => (self.newlines[j], j + 1),
+            Err(0) => (0, 0),
+            Err(j) if j == self.newlines.len() + 1 => (self.newlines[j - 1], j),
+            Err(j) => (self.newlines[j - 1], j),
+        };
+        let en = match self.newlines.binary_search(&span.end()) {
+            Ok(j) if j == self.newlines.len() + 1 => self.s.len(),
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs]: How can `j` ever be equal to `newlines.len() + 1`, if `binary_search` returns the index of the element when found? I assumed `j` is guaranteed to be `< newlines.len()`.

File: lrlex/src/lib/lexer.rs [hunk 2]
```diff
@@ -518,6 +508,31 @@ if 'IF'
...
+            s.push(vec!["a"; i].join(" "));
+        }
+        let s = s.join("\n");
+        let lexer = lexerdef.lexer(&s);
+        let lexemes = lexer.iter().map(|x| x.unwrap()).collect::<Vec<_>>();
+        assert_eq!(lexemes.len(), offs[70]);
+        assert_eq!(lexer.span_lines_str(Span::new(0, 0)), "");
+        assert_eq!(lexer.span_lines_str(Span::new(0, 2)), "\na");
+        assert_eq!(lexer.span_lines_str(Span::new(0, 4)), "\na\na a");
+        assert_eq!(lexer.span_lines_str(Span::new(0, 7)), "\na\na a\na a a");
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs:525]: Batman!

Comment (ltratt) [reply lrlex/src/lib/lexer.rs:525]: :p

File: lrlex/src/lib/lexer.rs [hunk 3]
```diff
@@ -310,30 +310,18 @@ impl<'lexer, 'input: 'lexer, StorageT: Copy + Eq + Hash + PrimInt + Unsigned>
...
-            }
-            (lexer.newlines[lexer.newlines.len() - 1], lexer.s.len())
-        }
-
-        let (st, _) = surrounding_line_off(self, span.start());
-        let (_, en) = surrounding_line_off(self, span.end());
+        let (st, _) = match self.newlines.binary_search(&span.start()) {
+            Ok(j) => (self.newlines[j], j + 1),
+            Err(0) => (0, 0),
+            Err(j) if j == self.newlines.len() + 1 => (self.newlines[j - 1], j),
```
Comment (ptersilie) [inline lrlex/src/lib/lexer.rs:316]: Is the same true for this line? `Err(j)` returns an index where an element can be inserted, so presumably the maximum value is insertion at the end, i.e. `newlines.len()`.

Comment (ltratt) [reply lrlex/src/lib/lexer.rs]: Oof, this is a classic "I'll copy and paste this without thinking" error. I've fixed this line *and* the original duff line in 7004af1.

**Comment by ptersilie** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**